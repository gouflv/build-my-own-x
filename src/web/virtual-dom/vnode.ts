export class Element<T extends keyof HTMLElementTagNameMap> {
  constructor(
    public tagName: T,
    public attr: Partial<HTMLElementTagNameMap[T]> | null | undefined,
    public children: (Element<any> | string)[]
  ) {}

  render() {
    const node = document.createElement(this.tagName)
    this.setNodeAttr(node)

    if (this.children.length) {
      this.children.forEach(child => {
        if (child instanceof Element) {
          node.appendChild(child.render())
        }
        if (typeof child === 'string') {
          node.appendChild(document.createTextNode(child))
        }
      })
    }

    return node
  }

  private setNodeAttr(node: HTMLElementTagNameMap[T]) {
    for (let key in this.attr) {
      if (!this.attr.hasOwnProperty(key)) {
        return
      }
      if (key === 'style') {
        node.style.cssText = this.attr[key] as any
      } else {
        node.setAttribute(key, this.attr[key] as any)
      }
    }
  }
}

export const createElement = <T extends keyof HTMLElementTagNameMap>(
  tagName: Element<T>['tagName'],
  attr?: Element<T>['attr'],
  children?: Element<T>['children']
) => {
  return new Element(tagName, attr || {}, children || [])
}
