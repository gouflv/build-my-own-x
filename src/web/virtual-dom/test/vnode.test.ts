import { createElement as h } from '../vnode'

describe('Test vnode', () => {
  const attr = {
    id: '#id',
    className: '.cls'
  }

  it('should create dom with attrs', () => {
    const el = h('div', attr)
    expect(el.tagName).toBe('div')
    expect(el.attr).toMatchObject(attr)
    expect(el.children).toStrictEqual([])
  })

  it('should create dom with children', () => {
    const el = h('ul', attr, [
      h('li', { id: '#c1', className: '.c1' }),
      h('li', { id: '#c2', className: '.c2' }),
      'isTextNode'
    ])
    expect(el.tagName).toBe('ul')
    expect(el.attr).toMatchObject(attr)
    expect(el.children.length).toBe(3)
    expect(el.children[0]).toMatchObject({
      tagName: 'li',
      attr: { id: '#c1', className: '.c1' }
    })
    expect(el.children[1]).toMatchObject({
      tagName: 'li',
      attr: { id: '#c2', className: '.c2' }
    })
    expect(el.children[2]).toBe('isTextNode')
  })

  it('should render dom node', () => {
    expect(h('div', attr).render()).toMatchSnapshot()
  })

  it('should render dom node with children', () => {
    expect(
      h('ul', attr, [
        h('li', { id: '#c1', className: '.c1' }),
        h('li', { id: '#c2', className: '.c2' }),
        h('li', null, ['isTextNode'])
      ]).render()
    ).toMatchSnapshot()
  })
})
