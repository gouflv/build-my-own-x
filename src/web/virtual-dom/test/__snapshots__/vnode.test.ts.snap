// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test vnode should render dom node 1`] = `
<div
  classname=".cls"
  id="#id"
/>
`;

exports[`Test vnode should render dom node with children 1`] = `
<ul
  classname=".cls"
  id="#id"
>
  <li
    classname=".c1"
    id="#c1"
  />
  <li
    classname=".c2"
    id="#c2"
  />
  <li>
    isTextNode
  </li>
</ul>
`;
